pkg/endpoint/endpoint.go

```diff
@@ -2474,6 +2474,12 @@ func (e *Endpoint) Delete(conf DeleteConfig) []error {
 
 	e.Stop()
 
+	// Wait for any pending endpoint regenerate() calls to finish. The
+	// latter bails out after taking the lock when it detects that the
+	// endpoint state is disconnecting.
+	e.buildMutex.Lock()
+	defer e.buildMutex.Unlock()
```

Comment:
Given the comment above `leaveLocked()` I'm surprised this was missing from this path. The lock ordering here looks correct compared to other usage of `buildMutex` (vs. `e.mutex`). It's hard to fully understand the impact of this with deletions potentially blocking more often depending on other endpoint creation/regeneration events going on in the system, but it does seem to be more correct behaviour.